freqtrade/misc.py

```diff
@@ -113,3 +113,21 @@ def format_ms_time(date: int) -> str:
     : epoch-string in ms
     """
     return datetime.fromtimestamp(date/1000.0).strftime('%Y-%m-%dT%H:%M:%S')
+
+
+def deep_merge_dicts(source, destination):
+    """
+    >>> a = { 'first' : { 'rows' : { 'pass' : 'dog', 'number' : '1' } } }
+    >>> b = { 'first' : { 'rows' : { 'fail' : 'cat', 'number' : '5' } } }
+    >>> merge(b, a) == { 'first' : { 'rows' : { 'pass' : 'dog', 'fail' : 'cat', 'number' : '5' } } }
+    True
+    """
+    for key, value in source.items():
```

Comment:
is this really necessary?
without testing it - wouldn't something like `source.update(destination)` also work?

i fear that combining the same keys (say "exchange") from multiple files will be VERY error prone, and if we do it this way, it'll need very good and clear documentation ...

Also it's inconsistent with your description, where "keys in later files override keys in earlier files" ... ;)

Reply:
> is this really necessary?
> without testing it - wouldn't something like `source.update(destination)` also work?
> 
> i fear that combining the same keys (say "exchange") from multiple files will be VERY error prone, and if we do it this way, it'll need very good and clear documentation ...
> 
> Also it's inconsistent with your description, where "keys in later files override keys in earlier files" ... ;)

Well, I implemented it first with dict.update() and found that in this case code do not merge options in a sabdict but substitute the whole subdict (like, in the example above, it takes 'exchange' subdict from the second file, without 'name' and other options since they are absent in the second file). So I defined that I need a kind of deep copy function. I'm not inventor of it, found it on stackoverflow.

It's not about hiding secrete keys in separate file with special file permissions. It also can be handy while working with multiple exchanges (or, at the other hand, working with multiple strategies on same exchange) -- this allows to separate parts of the config that differs between instances, leaving common part of configuration in a single common config file rather than copying config files and making errors committing changes to all of them...

It there an 'include' directive for json files? No, there is not. Only by self-made preprocessing...

In proposed implementation last config takes precedence, so an option in it overrides (overwrites) the option with the same name if it also appeared in all previous config files, specified with --config.

As an alternative, that can easily be applyed, the implementation may require unque location of every option, i.e. may generate an exception when it founds in the second config file an option that already appeared in one of preceding configs.

Reply:
I can (kind of) see the need - but I'm not sure what the benefit vs. complexity to support users to determine how the "finally" loaded config looks like .

How about saving the keys in a `.env` file and using [dotenv](https://pypi.org/project/python-dotenv/0.10.1/) to load specific values from there?

Reply:
I would not like personally add yet another format for storing configuration information -- dotenv implements plain config files with sections... Can be even more complecated for a user.

It's pretty easy to imagine how a resulting file can look like, imho -- it's a union of collections of options in all configs, where duplicates were removed, leaving value of a duplicated option set in the last config it appears.

As I said, for avoiding complexity, we can require each option be defined only once, in any of the config files specified -- so the example above should fail and one should use the following config.json along with keys.json:
```
{
...
    "exchange": {
        "name": "exchangename",
// no 'key', no 'secrete' here
        ...
    },
    ...
}
```
and key.json without ['exchange']['name'] respectively.

I understand that you are discussing alternatives, ways and options. Me too. It's all right.

Reply:
btw, i thought of printing the whole resulting config at the debug logging level (and did it for myself while developing the issue, indeed), but refused of that in the final committed version because of sensitive 'key' and 'secret' values. A function that hides them before logging should be added first...

Reply:
I'm not really concerned about you and me to understanding the logic of a merged dictionary - but of a "regular" trader without experience in thinking the way code works...

I think if it's documented properly it should be fine though...

Reply:
> I'm not really concerned about you and me to understanding the logic of a merged dictionary - but of a "regular" trader without experience in thinking the way code works...
I fully understand it, sure.

> I think if it's documented properly it should be fine though...
Think so as well.